### ISSUE TITLE
Webpack cannot handle string chunk.id

### ISSUE THREAD

**Issue by Truedrog** [webpack/webpack#3361]
**I'm submitting a bug report**
**Webpack version:**
2.1.0-beta.27

**Current behavior:**
For this snippet from my plugin:

```
compilation.plugin("before-chunk-ids", function(chunks) {
            chunks.forEach(function(chunk) {
                if(chunk.id === null) {
                    chunk.id = chunk.name;
                }
                if(!chunk.ids) {
                    chunk.ids = [chunk.id];
                }
            });
        });
```

where chunk.name is a string it being ignored by webpack when using require.ensure():
```
require.ensure([], (require) => {
                cb(null, require('./components/test/index').default);
            },"test");
```
Webpack generates this:
```
Promise.resolve().then((function (require) {
                cb(null, __webpack_require__("fd4e592df95c42900fdf").default);
            }).bind(null, __webpack_require__)).catch(__webpack_require__.oe);
```

**Expected/desired behavior:**
Not ignored chunk and generated output:

```
__webpack_require__.e/ nsure /("test").then((function (require) {
                cb(null, __webpack_require__("fd4e592df95c42900fdf").default);
            }).bind(null, __webpack_require__)).catch(__webpack_require__.oe);
```

**Comment by Kovensky** [comment]
I believe the issue is that your require uses a (completely) constant string, so no automatic async context is created, as webpack wouldn't know where to start (or end) the context.

I'm not really familiar with require.ensure, but maybe giving the module name as the first argument would make it async.

**Comment by Truedrog** [comment]
After some investigation i think the root of problem is in these parts:

webpack/lib/dependencies/DepBlockHelpers.js

```
DepBlockHelpers.getDepBlockPromise = function(depBlock, outputOptions, requestShortener, name) {
	if(depBlock.chunks) {
		var chunks = depBlock.chunks.filter(function(chunk) {
			return !chunk.hasRuntime() && typeof chunk.id === "number"; // This filters out my chunk so it's being ignored after.
		});
...
};
```

webpack/lib/JsonpMainTemplatePlugin.js

```
JsonpMainTemplatePlugin.prototype.apply = function(mainTemplate) {
	mainTemplate.plugin("local-vars", function(source, chunk) {
		if(chunk.chunks.length > 0) {
			return this.asString([
				source,
				"",
				"// objects to store loaded and loading chunks",
				"var installedChunks = {",
				this.indent(
					chunk.ids.map(function(id) {
						return id + ": 0"; // Here, my chunk id comes with some illegal characters like "test.1", so id should be quoted here, otherwise it's an error.
					}).join(",\n")
				),
				"};"
			]);
		}
		return source;
	});
...

}
```

**Comment by sokra** [comment]
That's doable. We've done the same change for module ids. It just need to fixed in many places.

There are many benefits of this. i. e. id = chunkhash or id = chunkname.
Feel free to try a PR. It's basically replacing `id` with `JSON.stringify(id)` and `typeof id === "number"` with `id !== null` and so on...

**Comment by Truedrog** [comment]
@sokra hello, unfortunately I am not so familiar with webpack codebase. 

Could you please clarify or point me where should replacing be done?

**Comment by Truedrog** [comment]
fixed by #4553
